project_euler/problem_54/sol1.py

```diff
@@ -0,0 +1,366 @@
+                self._cards.insert(0, ace_card)
+            return True
+        return False
+
+    def _is_straight(self) -> bool:
+        for i in range(4):
+            if self._cards[i + 1][0] - self._cards[i][0] != 1:
+                return False
+        return True
+
+    def _is_same_kind(self) -> int:
+        # Kind Values for internal use:
+        # 7: Four of a kind
+        # 6: Full house
+        # 3: Three of a kind
+        # 2: Two pairs
+        # 1: One pair
+        # 0: False
+        kind = val1 = val2 = 0
+        for i in range(4):
+            # Compare two cards at a time, if they are same increase 'kind',
+            # add the value of the card to val1, if it is repeating again we
+            # will add 2 to 'kind' as there are now 3 cards with same value.
+            # If we get card of different value than val1, we will do the same
+            # thing with val2
+            if self._cards[i][0] == self._cards[i + 1][0]:
+                if not val1:
+                    val1 = self._cards[i][0]
+                    kind += 1
+                elif val1 == self._cards[i][0]:
+                    kind += 2
+                elif not val2:
+                    val2 = self._cards[i][0]
+                    kind += 1
+                elif val2 == self._cards[i][0]:
+                    kind += 2
+        # For consistency in hand type (look at note in _get_hand_type function)
+        kind = kind + 2 if kind in [4, 5] else kind
+        # first meaning first pair to compare in 'compare_with'
+        first = max(val1, val2)
+        second = min(val1, val2)
+        # If it's full house (three count pair + two count pair), make sure
+        # first pair is three count and if not then switch them both.
+        if kind == 6 and self._card_values.count(first) != 3:
+            first, second = second, first
+        self._first_pair = first
+        self._second_pair = second
+        return kind
+
+    def _internal_state(self) -> List[Tuple[int, str]]:
+        # Internal representation of hand as a two tuple consisting of
+        # the value and suit of the card.
+        trans: dict = {"T": "10", "J": "11", "Q": "12", "K": "13", "A": "14"}
+        new_hand = self._hand.translate(str.maketrans(trans)).split()
+        final_hand = [(int(card[:-1]), card[-1]) for card in new_hand]
+        return sorted(final_hand)
+
+    def __repr__(self):
+        return f'{self.__class__}("{self._hand}")'
+
+    def __str__(self):
+        return self._hand
+
+    # Rich comparison operators (used in list.sort() and sorted() builtin functions)
+    # Note that this is not part of the problem but another extra feature where
+    # if you have a list of PokerHand objects, you can sort them just through
+    # the builtin functions.
+    def __eq__(self, other):
+        if isinstance(other, PokerHand):
+            return self.compare_with(other) == "Tie"
+        return NotImplemented
```

Comment:
Should this be NotImplementedError instead?

Reply:
No, I know people get confused between `NotImplemented` and `NotImplementedError` but those are two different things and not to be confused with each other. The major difference is that `NotImplementedError` is an exception while `NotImplemented` is a way to tell python that this method on those two objects is not implemented.

One of the ways `raise NotImplementedError` is used is when you make an abstract base class and you want the child classes to implement a method you defined otherwise, it will raise this error. For example:

```python
class Random(object):
    def some_method(self):
        raise NotImplementedError

class ChildRandom(Random):
    def some_method(self):
        return some_value
```
In the above example, if I didn't implement `some_method` then python will complain by raising `NotImplementedError`. You can achieve the same effect using `@abstractmethod` decorator.

`NotImplemented` is used when you define those dunder methods for the operators such as `+, -, ==, <` and the user inputs an object which is not the same as `self`, it will raise this error where it tells you `operator is not supported between the instances of <self class> and <other class>`
For example, if you're trying `a == b`, python will call `a.__eq__(b)` and if this returns `NotImplemented` then python will try `b.__eq__(a)` and if that one returns `NotImplemented` as well then python will raise an appropriate error.

https://docs.python.org/3.6/library/constants.html#NotImplemented
https://docs.python.org/3.6/library/exceptions.html#NotImplementedError